pkpdapp/pkpdapp/dash_apps/nca.py

```diff
@@ -0,0 +1,290 @@
+            self.C_0 = c0
+            self.time = np.insert(time, 0, 0, axis=0)
+            self.conc = np.insert(conc, 0, c0, axis=0)
+            self.DM = DM  # dose amount
+            self.administrationRoute = administrationRoute
+
+    def calculate_nca(self):
+        """
+        Calculates and stores all single dose NCA parameters
+        :return: None
+        """
+        self.AUC_0_last = self._AUC_0_last()
+        self.AUMC_0_last = self._AUMC_0_last()
+        self.Lambda_z, self.R2, self.Num_points = self._Lambda_z()
+
+        self.AUC_infinity = self._AUC_infinity()
+        self.AUC_infinity_dose = self._AUC_infinity_dose()
+        # self.AUCx_y
+        self.AUC_extrap_percent = self._AUC_extrap_percent()
+        self.CL = self._CL()
+        self.C_max, self.T_max = self._find_Cmax()
+        self.C_max_Dose = self._C_max_Dose()
+        # C_max_x_y
+
+        self.AUMC = self._AUMC()
+        self.AUMC_extrap_percent = self._AUMC_extrap_percent()
+        self.MRT = self._MRT()
+        self.Tlast = self._Tlast()  # ?? above LOQ?
+        self.T_half = np.log(2) / self.Lambda_z
+        # T_max_x_y
+        if self.administrationRoute == 'IVBolus':
+            self.V_ss = self._V_ss()
+        else:
+            self.V_ss = None
+
+        self.V_z = self._V_z()
+
+    @staticmethod
+    def extrapolate_c0(y, x):
+        """
+        Calculate initial concentration when not given, computed using a (log)
+        regression of the first two data points in a profile.
+        Arguments:
+        y {np.ndarray} -- y coordinates of points on curve
+        x {np.ndarray} -- x coordinates of points on curve
+
+        Returns:
+        c0 {double} -- extrapoloated c0
+        """
+        slope, intercept, r_value, _, _ = \
+            stats.linregress(x[:2], np.log(y[:2]), )
+        c0 = np.exp(intercept)
+        return c0
+
+    @staticmethod
+    def log_trapz(y, x):
+        """Calculate area under curve using log trapezoidal method (log mean).
+        Arguments:
+        :param y {np.ndarray} -- y coordinates of points on curve
+        :param x {np.ndarray} -- x coordinates of points on curve
+        Returns:
+        :return area {double} -- area under curve
+        """
+        y1 = y[:-1]  # exclude last term
+        x1 = x[:-1]  # exclude last term
+        y2 = y[1:]  # exclude first term
+        x2 = x[1:]  # exclude first term
+
+        area = np.sum((y1 - y2) / (np.log(y1) - np.log(y2)) * (x2 - x1))
+        return area
+
+    @staticmethod
+    def linlog_trapz(y, x, linlog=True):
+        """
+        Calculate area under curve using combination of linear and log
+        trapezoidal method (linear when increasing, log when decreasing) or
+        optionally just linear trapezoidal method
+        :param y {np.ndarray} -- y coordinates of points on curve
+        :param x {np.ndarray} -- x coordinates of points on curve
+        :param linlog {bool} -- use linear-log trapezoidal method
```

Comment:
If false, then what? Just linear?

Reply:
yes, looks like it from the doc-string (this code wasn't mine)